src/Mod/TechDraw/App/DrawViewDimension.cpp

```diff
@@ -574,107 +574,130 @@ std::string  DrawViewDimension::getFormatedValue(int partial)
-            } else {
-                repl = userVal;
+        // we can have 2 possible results:
+        // - the value in the base unit but without displayed unit
+        // - the value + unit (not necessarily the base unit!)
+        // the user can overwrite the decimal settings, so we must in every case use the formatSpecifier
+        // if useDecimals(), then formatSpecifier = global decimals, otherwise it is %.2f
+        QLocale loc;
+        double userVal;
+        bool checkDecimals = true;
+        if (showUnits() || (Type.isValue("Angle")) || (Type.isValue("Angle3Pt"))) {
+            formattedValue = qUserString; // result value + unit (not necessarily base unit!)
+            // remove unit
+            formattedValue.remove(rxUnits);
+            // to number
+            userVal = loc.toDouble(formattedValue);
+            if (userVal >= 1.0)
+                // we can assure we didn't make an error > 10% via getUserString()
+                checkDecimals = false;
+        }
+        if (checkDecimals){
+            // get value in the base unit with default decimals
+            // for the conversion we use the same method as in DlgUnitsCalculator::valueChanged
+            // at first remove space from unit string
+            QRegExp space(QString::fromUtf8("\\s"));
+            QString UnitString = qUserStringUnits;
+            UnitString.remove(space);
+            // if the unit is a metric lenght unit, nothing has to be converted
+            if ( UnitString == QString::fromUtf8("\xC2\xB5m") || UnitString == QString::fromLatin1("mm")
+                || UnitString == QString::fromLatin1("nm") || UnitString == QString::fromLatin1("m")
+                || UnitString == QString::fromLatin1("km") ) {
+                userVal = asQuantity.getValue();
             }
-        } else {
-            if (showUnits() || (Type.isValue("Angle")) || (Type.isValue("Angle3Pt"))) {
-                repl = formattedValue + qUserStringUnits;
-            } else {
-                repl = formattedValue;
+            else {
+                double convertValue = Base::Quantity::parse(QString::fromLatin1("1") + UnitString).getValue();
+                // we got now e.g. for "1 in" the value '25.4' because 1 in = 25.4 mm
+                // the result is now just val / convertValue because val is always in the base unit
+                userVal = asQuantity.getValue() / convertValue;
             }
         }
+        // we reformat the value
+#if QT_VERSION >= 0x050000
+        formattedValue = QString::asprintf(Base::Tools::toStdString(formatSpecifier).c_str(), userVal);
+#else
+        QString qs2;
+        formattedValue = qs2.sprintf(Base::Tools::toStdString(formatSpecifier).c_str(), userVal);
+#endif
+        // if abs(1 - userVal / formattedValue) > 0.1 we know that we make an error greater than 10%
+        // then we need more digits
+        if (userVal > 0.0 && abs(1 - formattedValue.toDouble() / userVal) > 0.1) {

```

Comment:
Is the intent really to isolate positive `userVal`, or to eliminate `userVal == 0.0`? If you are really just avoiding the div by zero, it would be clearer to state that, e.g.
```suggestion
        if (userVal != 0.0 && abs(1 - formattedValue.toDouble() / userVal) > 0.1) {
```
(I ask because I spent a disturbingly long time trying to figure out when `userVal` would be less than zero just now... 🤦)

Reply:
You have an expression of the form: abs(1 - x/y) > 0.1
This can be easily changed to the equivalent expression: abs(y - x) > 0.1 * y
This has two advantages:
* You don't have to check whether y is 0
* You avoid the rather expensive division by y

Reply:
> This can be easily changed to the equivalent expression: abs(y - x) > 0.1 * y

Thanks.